support: Add notion of primary admin.

ruchit2801 opened the pull request:
<!-- What's this PR for?  (Just a link to an issue is fine.) -->
This adds a field `is_primary_admin` in UserProfile model. This field gets updated for all users on migration, using a function `set_oldest_admin_primary_admin` in a migration file.  The logic I've used here is, get all `Realm` objects and for each of them, update `is_primary_admin` to be `True` for oldest admin, and `False` for rest. False is the default value in `is_primary_admin` model field.

TODO: 
Update this field on the creation of Realm

Fixes : #12249

ruchit2801 commented:
@rishig Can you review this, I assume the script in `set_oldest_user_primary_admin` is correct.

ruchit2801 commented:
Thanks! @YJDave

ruchit2801 commented on zerver/migrations/0219_userprofile_is_primary_admin.py:
```diff
@@ -0,0 +1,33 @@
+    UserProfile = apps.get_model('zerver', 'Realm')
+    for realm in Realm.objects.all():
+        realm_admins = UserProfile.objects.filter(realm=realm, is_realm_admin=True)
+        oldest_realm_admin = realm_admins.order_by('-date').first()
```
This would be `order_by('-date_joined')`, instead of `date`, I'll correct it soon :)

timabbott commented:
This change isn't mergable without code in the `do_create_user` flow for realm creation to set the flag; otherwise it'll be missing on new rows.

I think we could reasonably take this opportunity to do the messy migration to a `role` field with values GUEST/MEMBER/ADMIN/PRIMARY ADMIN, and move things like `.is_realm_admin` to be properties that do a comparison on those values.  But it's probably 20x the work to do that (because we'll need to think about how what to put for bots, and whether it's valid to have a bot that's an admin, etc.), so my preference is we just implement this with the data model in this PR.

(I think this is unrelated to our discussion of "new member" for the reasons of the best way to calculate we discussed earlier today)

ruchit2801 commented:
> This change isn't mergable without code in the do_create_user flow for realm creation to set the flag; otherwise it'll be missing on new rows. @ruchit2801 can you work on that? We should also adjust our existing tests that create a new realm to have them check that this field is set, and probably the test for inviting a new user as an admin to check this field doesn't get set.

Yepp! This is on my todo list. I'll push it today or latest by tonight :)

ruchit2801 commented:
>We should also adjust our existing tests that create a new realm to have them check that this field is set

@timabbott, regarding this, I could not figure out where realm creation is being tested.  function `test_ldap_registration_multiple_realms(self) -> None:` in `test_signup.py`, seems to test realm creation. There are some other places in tests where `do_create_realm` is used, but those are not for testing realm creation.

ruchit2801 commented:
Hey! @timabbott. I pushed code to update `is_primary_admin` upon realm creation. Can you check this out? Still waiting for your suggestion on updating tests :)

timabbott commented:
@ruchit2801 you should be looking at the end-to-end tests; we generally don't test functions like `do_create_realm` directly, but instead test the end-to-end registration flow for creating an organization.  There's a lot of tests with `create_realm` in their name in `zerver/tests/test_signup.py` for this purpose.

ruchit2801 commented:
@timabbott , for testing, in `test_signup.py`, I've written tests in `check_able_to_create_realm` to verify that a flag `is_primary_admin` is set whenever a new realm is created. In addition to this, I've also added tests in the same file in `InviteUserTest` class, to verify that this flag is not set when any other user is invited as admin, guest or etc.
So, I imagine this is complete now.

timabbott commented:
@ruchit2801 I pushed back changes to this PR to fix a few issues:
* Weird Python style with extra parens.
* An unnecessary database query editing the user.

I don't think this is ready to merge, though, because we also need to handle the happens when e.g. the primary admins stops being an admin; otherwise, this state will end up broken resulting in manual migrations in the future.  I think the right protocol for that is:
* The primary admin can transfer primary admin status to another user.
* Admins can't remove admin status from the primary admin; they must first have the primary admin status transferred.
* A primary admin user cannot be deactivated (I think this basically ends up replacing the existing "cannot deactivate the last administrator" logic).

This ensures that every organization will always have an active primary admin user.

@rishig thoughts?

rishig commented:
yeah, that sounds good to me.

ruchit2801 commented:
@timabbott, thanks for the reasonable changes in the PR. 
> The primary admin can transfer primary admin status to another user.
Admins can't remove admin status from the primary admin; they must first have the primary admin status transferred.
A primary admin user cannot be deactivated (I think this basically ends up replacing the existing "cannot deactivate the last administrator" logic).

Alright, so I'll focus on this part next. This seems to be a bit more work. I think I'll complete this is a separate PR.

timabbott commented:
No need for a separate PR -- just do them in additional commits in this
PR.  It's better to keep the discussion in one place for stacked worked
like this.

On Tue, May 14, 2019 at 8:22 PM ruchit2801 <notifications@github.com> wrote:

> @timabbott <https://github.com/timabbott>, thanks for the reasonable
> changes in the PR.
>
> The primary admin can transfer primary admin status to another user.
> Admins can't remove admin status from the primary admin; they must first
> have the primary admin status transferred.
> A primary admin user cannot be deactivated (I think this basically ends up
> replacing the existing "cannot deactivate the last administrator" logic).
>
> Alright, so I'll focus on this part next. This seems to be a bit more
> work. I think I'll complete this is a separate PR.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/zulip/zulip/pull/12298?email_source=notifications&email_token=AAU6NWVKCQHTXHMPD2JIWKTPVN6YRA5CNFSM4HLPHAMKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODVNMPQY#issuecomment-492488643>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAU6NWXCQWPFB2GCTWKSKV3PVN6YRANCNFSM4HLPHAMA>
> .
>

ruchit2801 commented:
Hey @timabbott! So this is update in this PR.
 * Added a code to stop other admins from deactivating and changing the admin status of a primary admin. 
* Whenever a primary admin decides to transfer this status to some other user, we set `is_admin` and `is_primary_admin` fields for that user and remove `is_primary_admin`(set it to `False`) field of current primary admin (Current primary admin will lose the primary admin status upon transfer).  
* A transfer of primary_admin status to some other user may cause a broken `is_primary_admin` field upon migration. So, I've added logic in the migration file to handle this. This logic will set the oldest user as the primary admin only when there are no primary admins in the organization.

ruchit2801 commented:
For the UI part, I think we should display `primary admin` option in User roles, which would be only visible to primary admin.

rishig commented:
> Whenever a primary admin decides to transfer this status to some other user, we set is_admin and is_primary_admin fields for that user and remove is_primary_admin

From a policy perspective, I think it's fine/preferable to allow multiple primary owners. So we don't need to remove the `is_primary_admin` designation when the primary admin promotes someone else.

rishig commented:
> For the UI part, I think we should display primary admin option in User roles, which would be only visible to primary admin.

You mean, visible to primary admins and not to admins? Yeah, that makes sense.

timabbott commented:
@ruchit it appears I must hvae accidentally force-pushed another branch to this PR; sorry about that!  Can you reset this PR to your code from 9416d4a86a9283ae4f8b86c81d9447ee63e91f47?

ruchit2801 commented:
Sure! no worries :)

timabbott commented:
@ruchit2801 looks like you haven't done so yet?

ruchit2801 commented:
I was AFK for the past two days. Changes pushed @timabbott. Ready for review.

ruchit2801 commented:
@timabbott Looks like you missed to review this.

timabbott commented:
We'll want to change the strings with something like this; @rishig have we settled on a final user-facing name for this?  If it's going to be "organization owner" or something, now's the time to do that.

```
diff --git a/zerver/views/users.py b/zerver/views/users.py
index f557e63473..f7250afa4c 100644
--- a/zerver/views/users.py
+++ b/zerver/views/users.py
@@ -43,7 +43,7 @@ def deactivate_user_backend(request: HttpRequest, user_profile: UserProfile,
                             user_id: int) -> HttpResponse:
     target = access_user_by_id(user_profile, user_id)
     if target.is_primary_admin:
-        return json_error(_('Cannot deactivate the primary admin of organization'))
+        return json_error(_('Cannot deactivate the only primary administrator'))

     if check_last_admin(target):
         return json_error(_('Cannot deactivate the only organization administrator'))
     return _deactivate_user_profile_backend(request, user_profile, target)
@@ -102,7 +102,7 @@ def update_user_backend(request: HttpRequest, user_profile: UserProfile, user_id
 
     if is_admin is not None and target.is_realm_admin != is_admin:
         if not is_admin and target.is_primary_admin:
-            return json_error(_('Cannot change admin status of primary admin'))
+            return json_error(_('Only a primary administrator can remove primary administrator status'))
 
     if is_admin is not None and target.is_realm_admin != is_admin:
         if not is_admin and check_last_admin(user_profile):
```
(Because we generally don't spell it "admin" in the UI).

The API changes here so far look reasonable, but this PR needs more work:
* We need to change the "deactivate myself" code path to prevent the only primary administrator from deactivating themself
* We need to display "Primary administrator"  (or whatever the string is going to be) in the frontend UI places where we display a user's role in the UI so users understand what's going on (profiles, "display settings", and the "organization users" table).  
* the other pieces above (https://github.com/zulip/zulip/pull/12298#issuecomment-492377708), e.g. making the UI for changing a user's role include a "primary administrator" option.

timabbott commented on zerver/tests/test_users.py:
```diff
@@ -67,6 +68,21 @@ def test_do_change_is_admin(self) -> None:
+        hamlet = self.example_user('hamlet')
+
+        self.assertEqual(hamlet.is_primary_admin, False)
+        do_change_is_primary_admin(hamlet, True)
```
A few notes on this test:
(1) We should change `populate_db` to make sure Iago is set as the primary administrator in the prebuilt test database.
(2) It'd generally better for testing to hit the API, rather than calling `do_change_...` functions, since then we're testing more of the logic.  So the call on this line should do that.

timabbott commented on zerver/tests/test_users.py:
```diff
@@ -67,6 +68,21 @@ def test_do_change_is_admin(self) -> None:
+        req = dict(is_primary_admin=ujson.dumps(True))
+        self.assertFalse(iago.is_primary_admin)
+        result = self.client_patch('/json/users/{}'.format(iago.id), req)
+        self.assert_json_success(result)
```
Also, a test like this should never end with just checking success -- we should also refetch Iago from the database and verify his state after it was supposedly changed in an API call.

rishig commented:
> From a policy perspective, I think it's fine/preferable to allow multiple primary owners. So we don't need to remove the is_primary_admin designation when the primary admin promotes someone else.

After looking into this more, it's possible it'll make things like GDPR compliance more complicated. So let's just go with "there can only be exactly one primary admin" for now. Sorry about the change here!

For the strings, let's go with "Primary administrator". Thanks!

ruchit2801 commented:
To support exactly one primary admin, 
* The logic where we warn the user about deactivating the last administrator of the realm will be removed since we are not allowing the primary admin to remove himself (and to do so, he need to transfer his primary admin status, so at any time, there must be at least one admin (the primary admin if there is only one). 
* There must be exactly one primary admin, so if he tries to deactivate himself or change his role, we will warn him that he should first transfer his primary admin status. 
* If he decides to transfer his primary_admin role, we will warn him about the loss of his own primary_admin status.  

I thought the following design for this. 
* Implement the `do_change_primary_admin` in actions.py which will take two arguments, one target user and one user_profile, and it will then transfer the primary admin status from user_profile to target user. 
* @rishig, One of my doubt is: When I did the migrations, I realized that `Realm.objects.all()` in a migration file, returns an empty QuerySet (and hence, iago is not changed to primary_admin on migration). I don't know all logic about how migration updates the database of the development version (maybe populate_db is doing it?) but iago is not set as primary_admin on doing the migration. Tim asked set iago as primary admin in populate_db and since there is no any primary admin when we want to do this, so should I just pass iago as both target and user_profile(that would work by the way)?

rishig commented:
cool, that design seems reasonabel for do_change_primary_admin. 

I wouldn't worry about how the migratoins interact with the dev environment; you can just set Iago to be a primary admin by changing `populate_db.py`.

ruchit2801 commented:
- [x] Prevent primary admin from deactivating himself (deactivate myself codepath)
- [x] Only one primary admin per realm allowed
- [ ]  UI part (display `primary admin` in user role form of all users)
- [x] Change the message strings as Tim suggested
- [x] improved the tests from the last changes pushed  

Some parts of this PR, which I think are the flags which should be mentioned : 
After this update, the last admin and the last user (if in any case), both of them must be a primary admin of the organization. 
So, I've removed the code paths which included warning these users and added a common primary admin code for all cases. 
@rishig or @timabbott can one of you review the backend part? I'll start working on the UI part tomorrow, good night for now :)
